pkg/cmd/infra/router/template.go

```diff
@@ -565,3 +566,88 @@ func (o *TemplateRouterOptions) blueprintRoutes(routeclient *routeclientset.Clie
 
 	return blueprints, nil
 }
+
+// makeTLSConfig checks whether metrics TLS is configured and
+// if so returns a tls.Config whose certificate is automatically
+// reloaded every 5 minutes or in response to file changes.
+func makeTLSConfig() (*tls.Config, error) {
+	certFile := env("ROUTER_METRICS_TLS_CERT_FILE", "")
+	if len(certFile) == 0 {
+		return nil, nil
+	}
+	keyFile := env("ROUTER_METRICS_TLS_KEY_FILE", "")
+
+	// Make sure we have something useful to start with
+	certificate, err := tls.LoadX509KeyPair(certFile, keyFile)
+	if err != nil {
+		return nil, err
+	}
+
+	// Safely reload the certificate as requested.
+	var lock sync.Mutex
+	reload := make(chan bool) // value will be false for an unscheduled reload
+	go func() {
+		for {
+			select {
+			case scheduled := <-reload:
+				latest, err := tls.LoadX509KeyPair(certFile, keyFile)
+				if err != nil {
+					// TODO: add a prometheus metric? Fail a health check? This should surface somehow...
+					log.Error(err, "failed to reload certificate", "cert", certFile, "key", keyFile)
+					break
+				}
+				lock.Lock()
+				certificate = latest
+				lock.Unlock()
+				// TODO: add a prometheus metric instead?
+				if scheduled {
+					log.V(4).Info("reloaded certificate on fixed schedule", "cert", certFile, "key", keyFile)
+				} else {
+					log.V(0).Info("reloaded modified certificate", "cert", certFile, "key", keyFile)
+				}
+			}
+		}
+	}()
+
+	// Reload when the files change or when the fixed timer fires
+	ticker := time.NewTicker(5 * time.Minute)
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	if err := watcher.Add(certFile); err != nil {
+		return nil, err
+	}
+	if err := watcher.Add(keyFile); err != nil {
+		return nil, err
+	}
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				reload <- true
+			case _, ok := <-watcher.Events:
+				if ok {
+					// Ignoring the event type for now to avoid any portability
+					// issues at the expense of potentially spurious reloads on
+					// irrelevant events (which should be rare, this is a mounted
+					// secret).
```

Comment:
What you're like to see is a flurry of events for the same file (e.g., a `CREATE` event and a `CHMOD` event for the same update).

Reply:
Disambiguating means trusting the event type, and I have reason to doubt they are reliable across contexts based on some previous experience with the lib. Given how infrequent the updates should be (when a cert is rotated), do we really care if we trigger a few spurious reloads in this case?

Reply:
And by "reload" here I just mean we'll read the disk contents of the cert/key file (which are probably mounted through a RAM volume anyway)

Reply:
Agreed, it's not terrible to trigger spurious reloads.